### PR TITLE
Improve messages due `VertexException` in `SequentialPrimaryVertexFitterAdapter` and clean-up of `BeamSpotOnlineProducer`

### PR DESCRIPTION
**Pull request by mmusich**
#### PR description:

This is a follow-up to https://github.com/cms-sw/cmssw/issues/45555#issuecomment-2515124788.

   * a4a0d1685ad4716ef65a472445fd755ea42e7a3d annotates `SequentialPrimaryVertexFitterAdapter::fit` exceptions of the `VertexExcpetion` type according to the suggestions at https://github.com/cms-sw/cmssw/issues/45555#issuecomment-2515124788
   * a390d6b643460a04903e7acc49a4e997495c0c8b contains a generalized clean-up / refactoring of `BeamSpotOnlineProducer` in order to be more modular (and hopefully more easily debuggable).

#### PR validation:

`cmssw` compiles. Run an HLT job in normal conditions and didn't observe anomalies.
Inducing the type of `VertexException` that is being caught now, I get the following type of log messages:

```
%MSG-e UnusableBeamSpot:  PrimaryVertexProducer:hltVerticesPF  07-Dec-2024 11:13:50 CET Run: 386614 Event: 48610345
Beamspot with invalid errors [            0           0           0
             0           0           0
             0           0           0 ]
%MSG
----- Begin Fatal Exception 07-Dec-2024 11:13:50 CET-----------------------
An exception of category 'VertexException' occurred while
   [0] Processing  Event run: 386614 lumi: 55 event: 48610345 stream: 0
   [1] Running path 'MC_AK8PFJetPNet_v4'
   [2] Calling method for module PrimaryVertexProducer/'hltVerticesPF'
   [3] While processing SequentialPrimaryVertexFitterAdapter::fit() with BeamSpot parameters: 
-----------------------------------------------------
              Beam Spot Data

 Beam type    = -1
       X0     = 0 +/- 0 [cm]
       Y0     = 0 +/- 0 [cm]
       Z0     = 0 +/- 0 [cm]
 Sigma Z0     = 0 +/- 0 [cm]
 dxdz         = 0 +/- 0 [radians]
 dydz         = 0 +/- 0 [radians]
 Beam width X = 0 +/- 0 [cm]
 Beam width Y = 0 +/- 0 [cm]
 EmittanceX   = 0 [cm]
 EmittanceY   = 0 [cm]
 beta-star    = 0 [cm]
-----------------------------------------------------


Exception Message:
BasicSingleVertexState::could not invert error matrix 
```

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

N/A

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc [hunk 1]
```diff
@@ -88,151 +98,161 @@ void BeamSpotOnlineProducer::fillDescriptions(edm::ConfigurationDescriptions& iD
...
   }
-  if (fallBackToDB) {
-    edm::ESHandle<BeamSpotObjects> beamhandle = iSetup.getHandle(beamToken_);
-    const BeamSpotObjects* spotDB = beamhandle.product();
 
-    // translate from BeamSpotObjects to reco::BeamSpot
-    reco::BeamSpot::Point apoint(spotDB->x(), spotDB->y(), spotDB->z());
+  double sigmaZ = (theSetSigmaZ > 0) ? theSetSigmaZ : spotDB.sigmaZ();
+  result = reco::BeamSpot(apoint, sigmaZ, spotDB.dxdz(), spotDB.dydz(), spotDB.beamWidthX(), matrix);
+  result.setBeamWidthY(spotDB.beamWidthY());
```
[the remark this answers is not in the document]
Comment (mmusich) [reply RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc:163]: > but why not also `setBeamWidthX`

the x-width is already set in the constructor at L162:
```C++
  result = reco::BeamSpot(apoint, sigmaZ, spotDB.dxdz(), spotDB.dydz(), spotDB.beamWidthX(), matrix);
```
as per method signature:

https://github.com/cms-sw/cmssw/blob/e9b58bef8b37e6113ba31a03429ffc4c300adb12/DataFormats/BeamSpot/interface/BeamSpot.h#L39-L45

as both widths are constructed with the same value (in the hypothesis of round-optics)

https://github.com/cms-sw/cmssw/blob/e9b58bef8b37e6113ba31a03429ffc4c300adb12/DataFormats/BeamSpot/interface/BeamSpot.h#L50-L51

setting explicitly the y component is necessary if the two differ (and in principle they will differ in 2025, if the flat optics is used).

**Comment by jfernan2** [comment]
+1

File: RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (perrotta) [reply RecoVertex/BeamSpotProducer/plugins/BeamSpotOnlineProducer.cc:163]: > setting explicitly the y component is necessary if the two differ (and in principle they will differ in 2025, if the flat optics is used).

Ah, ok: that was indeed my worry. Thank you for confirming that the two directions can be set independently,

**Comment by perrotta** [comment]
+alca

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**